torchtune/datasets/_packed.py

```diff
@@ -0,0 +1,163 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Dict, List, Optional
+
+import torch
+
+from torch.utils.data import Dataset
+from torchtune.utils import get_world_size_and_rank
+from torchtune.utils.collate import _padded_collate_packed
+from tqdm import tqdm
+
+
+class PackedDataset(Dataset):
+    """
+    Performs greedy sample packing on a provided dataset. This is done as a single
+    preprocessing step before training begins. Shuffling is done outside of this
+    class on packed samples with a ``Sampler`` as part of the dataloader. Currently,
+    this only supports in-memory datasets.
+
+    The class loads, tokenizes, and packs examples on initialization - no tokenization is done during training.
+
+    The general flow on initialization is: load tokenized sample -> add to buffer ->
+        when buffer is long enough, add to self.packs.
+
+    During training, returns self.packs[idx] as input, label, attention mask, and
+    position ids. The attention mask is a lower triangular block mask to prevent
+    samples from cross-attending within a pack. The position ids indicate the position
+    of each token relative to its sample within a pack. These are all padded to max
+    sequence length, so a batch-wise collator is not needed.
```

Comment:
Overall great docstring. There's a lot going on here, so wonder if it might make sense to just write out an explicit example here (especially what mask and input_pos look like when packing a couple samples). It'd be really nice to have something like that in the live docs for people to refer to.